src/Symfony/Component/HttpClient/Response/AsyncResponse.php

```diff
@@ -0,0 +1,355 @@
+                        unset($asyncMap[$response]);
+                    }
+
+                    yield $r => $chunk;
+                    continue;
+                }
+
+                $context = new AsyncContext($r->passthru, $r->client, $r->response, $r->info, $r->content, $r->offset);
+                if (null === $stream = ($r->passthru)($chunk, $context)) {
+                    if ($r->response === $response && (null !== $chunk->getError() || $chunk->isLast())) {
+                        throw new \LogicException('A chunk passthru cannot swallow the last chunk.');
+                    }
+
+                    continue;
+                }
+                $chunk = null;
+
+                if (!$stream instanceof \Iterator) {
+                    throw new \LogicException(sprintf('A chunk passthru must return an "Iterator", "%s" returned.', get_debug_type($stream)));
+                }
+
+                while (true) {
+                    try {
+                        if (null !== $chunk) {
+                            $stream->next();
+                        }
+
+                        if (!$stream->valid()) {
+                            break;
+                        }
+                    } catch (\Throwable $e) {
+                        $r->info['error'] = $e->getMessage();
+                        $r->response->cancel();
+
+                        yield $r => $chunk = new ErrorChunk($r->offset, $e);
+                        $chunk->didThrow() ?: $chunk->getContent();
+                        unset($asyncMap[$response]);
+                        break;
+                    }
+
+                    $chunk = $stream->current();
+
+                    if (!$chunk instanceof ChunkInterface) {
+                        throw new \LogicException(sprintf('A chunk passthru must yield instances of "%s", "%s" yielded.', ChunkInterface::class, get_debug_type($chunk)));
+                    }
+
+                    if (null !== $chunk->getError()) {
+                        // no-op
+                    } elseif ($chunk->isFirst()) {
+                        $e = $r->openBuffer();
+
+                        yield $r => $chunk;
+
+                        if (null === $e) {
+                            continue;
+                        }
+
+                        $r->response->cancel();
+                        $chunk = new ErrorChunk($r->offset, $e);
+                    } elseif ('' !== $content = $chunk->getContent()) {
+                        if (null !== $r->shouldBuffer) {
+                            throw new \LogicException('A chunk passthru must yield an "isFirst()" chunk before any content chunk.');
+                        }
+
+                        if (null !== $r->content && \strlen($content) !== fwrite($r->content, $content)) {
+                            $chunk = new ErrorChunk($r->offset, new TransportException(sprintf('Failed writing %d bytes to the response buffer.', \strlen($content))));
```

Comment:
Message is misleading. It says it failed to write xy bytes, while code ignores case when fwrite partially succeeds. As an example, assuming $content has length 4096, message might say it failed to write 4096 bytes, when in reality fwrite might succeeded writing first 1000 bytes and that's what it will return. Message should use diff between fwrite result and strlen($content) in this message.

Reply:
We don't care: nobody counts the exact bytes in error messages. Also because this should be exceptional, and this borrows from already existing code in (Transport)ResponseTrait. PR welcome on 4.4 if you think we should care.